api/src/main/java/org/apache/iceberg/expressions/Evaluator.java

```diff
@@ -151,5 +151,10 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
     public <T> Boolean startsWith(Bound<T> valueExpr, Literal<T> lit) {
       return ((String) valueExpr.eval(struct)).startsWith((String) lit.value());
     }
+
+    @Override
+    public <T> Boolean notStartsWith(Bound<T> valueExpr, Literal<T> lit) {
+      return !((String) valueExpr.eval(struct)).startsWith((String) lit.value());
```

Comment:
we could just run !startsWith()

Reply:
I agree here. `!startsWith()` is good because it avoids implementing the same thing twice.

Reply:
Ok. That sounds good to me. As I mentioned above to @shangxinli there are a few places where !startsWith can be substituted.

As I see it currently, they are `ResidualEvaluator` and `BoundLiteralPredicate`. Should I do the same thing there?

Reply:
I've fixed the usage in this evaluator, please let me know if you think I should do the same in `ResidualEvaluator` and `BoundLiteralPredicate`.

Reply:
Update

For `ResidualEvaluator`, it can't be `!startsWith(ref, lit)` as `ResidualEvaluator` returns an expression, so it would have to be `startsWith(ref, lit).negate()`. This works since we're returning `alwaysTrue()` or `alwaysFalse()`, but it is a little confusing. Also, in the same file, we don't define `notIn` by using `in` in this way. I personally think `ResidualEvaluator` should be left as is.

Additionally, `BoundLiteralPredicate` cannot be defined in terms of `STARTS_WITH` as the function signature is just `test(T value)` and then the operation is caught in a switch statement and the operation is stored as a value on `this` (the `BoundLiteralPredicate` instance). So we'd have to return `negate().test(value)` which is also pretty confusing and likely error prone.

I'm going to close this conversation as it seems that the instances where we can easily and simply replace `notStartsWith` with `!starsWith` have all been covered. Feel free to reopen it or comment further if you disagree.